src/Mvc/Mvc.Core/src/HotReload/HotReloadService.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection.Metadata;
+using System.Threading;
+using Microsoft.AspNetCore.Mvc.Controllers;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Metadata;
+using Microsoft.Extensions.Primitives;
+
+[assembly: MetadataUpdateHandler(typeof(Microsoft.AspNetCore.Mvc.HotReload.HotReloadService))]
+
+namespace Microsoft.AspNetCore.Mvc.HotReload
+{
+    internal sealed class HotReloadService : IActionDescriptorChangeProvider, IDisposable
+    {
+        private readonly DefaultModelMetadataProvider? _modelMetadataProvider;
+        private readonly DefaultControllerPropertyActivator? _controllerPropertyActivator;
+        private CancellationTokenSource _tokenSource = new();
+
+        public HotReloadService(
+            IModelMetadataProvider modelMetadataProvider,
+            IControllerPropertyActivator controllerPropertyActivator)
+        {
+            ClearCacheEvent += NotifyClearCache;
+            if (modelMetadataProvider.GetType() == typeof(DefaultModelMetadataProvider))
+            {
+                _modelMetadataProvider = (DefaultModelMetadataProvider)modelMetadataProvider;
+            }
+
+            if (controllerPropertyActivator is DefaultControllerPropertyActivator defaultControllerPropertyActivator)
```

Comment:
This type is sealed?

Reply:
Yup. internal sealed. An alternative to doing this is to put the handler in the type itself, but I wanted to try and contain all of the clearing bits in one place particularly because there's indirection involved turning a static method call in to clearing per-instance caches.